rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+
+
+# Detailed design
+
+
+## High level architecture
+
+Below diagram shows the proposed architecture of the URL Service.
+
+![URL Service architecture](../images/url_service/new_architecture.png)
+
+
+## Plugin contracts
+
+The is to provide developers the same experience on the server and browser.
+
+Below are preliminary interfaces of the new URL Service. `IUrlService` will be
+a shared interface defined in `/common` folder shared across server and browser.
+This will allow us to provide users a common API interface on the server and
+browser, wherever they choose to use the URL Service:
+
+```ts
+/**
+ * Common URL Service client interface for the server-side and the client-side.
+ */
+interface IUrlService {
+  generators: IUrlGeneratorClient;
+  shortUrls: IShortUrlClient;
+}
+```
+
+The URL generator client will basically be a registry of URL generators:
+
+```ts
+/**
+ * URL generator registry.
+ */
+interface IUrlGeneratorClient {
+  register<S>(urlGenerator: UrlGeneratorDefinition<S>): UrlGenerator;
+  get<S>(id: string): UrlGenerator<S>;
+}
+
+interface UrlGeneratorDefinition<State> {
+  id: string;
+  createUrl: (state: State) => Promise<string>;
+  isDeprecated?: boolean;
+  migrate?: (state: State) => Promise<{ state: unknown; id: string }>;
+}
+
+interface UrlGenerator<State> {
+  isDeprecated: boolean;
+  createUrl(state: State): Promise<URL>;
+  migrate(): Promise<never | { state: unknown; id: string }>;
+}
+```
+
+The short URL client `IShortUrlClient` which will be the same on the server and
+browser. However, the server and browser might add extra utility methods for
+convenience.
+
+
+```ts
+/**
+ * CRUD-like API for short URLs.
+ */
+interface IShortUrlClient {
+  /**
+   * Create a new short URL from URL generator.
+   *
+   * @param urlGeneratorId ID of the URL generator.
+   * @param state State object to provide to the URL generator.
+   * @param slug Optionally, provide a custom slug (ID) of the short URL.
+   */
+  create(
+    urlGeneratorId: string,
+    state: unknown,
+    slug?: string
+  ): Promise<{ slug: string; url: URL }>;
+
+  /**
+   * Delete a short URL.
+   * 
+   * @param slug The slug (ID) of the short URL.
+   * @return Returns true if deletion was successful.
+   */
+  delete(slug: string): Promise<boolean>;
+
+  /**
+   * Fetch short URL.
+   * 
+   * @param slug The slug (ID) of the short URL.
+   */
+  get(slug: string): Promise<{ slug: string; shortUrl: object; url: URL }>;
+
+  /**
+   * Same as `get()` but it also increments the "view" counter and the
+   * "last view" timestamp of this short URL.
+   * 
+   * @param slug The slug (ID) of the short URL.
+   */
+  resolve(slug: string): Promise<{ slug: string; shortUrl: object; url: URL }>;
+}
+```
+
+Note, that in the new service to generate a short URL the developer will have to
+use a URL generator.
+
+```ts
+const shortUrl = await share.url.shortUrls.create({
+  generator: 'DISCOVER_DEEP_LINKS',
+  state: {
```

Comment:
how does the consumer know about this state ?|
- are we hoping to make this a very small set of settings the user could set ?
- or is the plan to have full state in here ?

maybe we need to separate the two ? if discover app wants to generate the short url, it has full knowledge of state. In this case i am not sure we should call this url generators at all. 
- discover could just use short url service. the short url saved object should contain appId and appState, and the endpoint should redirect with core.app.navigateWithState(appId, appState)

when someone external however wants to navigate to discover, they don't know the full state, but ight know some information. In this case the generators make sense. Even here they probably shouldn't be called url generators, but state generators. i want to navigate to discover and select an index pattern .....
- discoverPlugin.navigate({ indexPattern }) maybe just a method on discover plugin for that ?
- or the state generator, which takes in this simple object and returns full state i can then use with core.app.navigateTo ?

Reply:
i am not sure i fully follow ? if we would to make filters part of that public API .... we need to make filter structure public, or provide some other public API to generate this filters (which we already have)

Reply:
Agreed with Peter to rename "URL generators" to "locators". Also see the latest proposed API for creating short URLs.

```ts
const shortUrl = await plugins.discover.locator.createShortUrl(
  {
    indexPattern: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
    highlightedField: 'foo',
  }
  'human-readable-slug',
});
```